### PR TITLE
Create linearisation of AtmosModel

### PR DESCRIPTION
**Pull request by simonbyrne**
@mwarusz This is an alternative approach to #443, based on using wrappers around `AtmosModel` to operate on subcomponents. Would this be easier to work with?

### CODE REVIEW

**Comment by tapios** [comment]
Looks very nice and clean!

**Comment by tapios** [comment]
We'll primarily need the linear components in the vertical (for 1D IMEX), so only vertical fluxes. An option may be just to implement this, rather than the 3D version here. Or else, if we want both, would it be useful to think of a good abstraction of doing that, rather than defining separate 1d linear, 3d linear models, each with different options for including or excluding gravity waves?

**Comment by lcw** [comment]
If you look at how 1D IMEX worked in the previous framework, https://github.com/climate-machine/CLIMA/pull/373, there is one linear model for 1D and 3D.  We did this by specifying the 3D linearization and zeroing out horizontal derivatives in the 1D computational kernels.

**Comment by simonbyrne** [comment]
bors try

File: test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+               Rusanov(),
+               CentralNumericalFluxDiffusive(),
+               CentralGradPenalty())
+
+  linmodel = LinearType(model)
+  lindg = DGModel(linmodel,
+               grid,
+               Rusanov(),
+               CentralNumericalFluxDiffusive(),
+               CentralGradPenalty(); auxstate=dg.auxstate)
```
Comment (simonbyrne) [inline test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl]: Note we need to share the aux state from the main model so that it has access to the reference state. It may make sense to move the reference state to the model instead (in which case it would then have its own aux state).

**Comment by simonbyrne** [comment]
bors try

File: src/Atmos/Model/linear.jl [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+init_state!(lm::AtmosLinearModel, state::Vars, aux::Vars, coords, t) = nothing
+
+
+function linear_pressure(moisture, orientation, state::Vars, aux::Vars)
+  ρ = state.ρ
+  invρ = inv(ρ)
+  TS = thermo_state(moisture, orientation, state, aux)
+  # TODO: avoid dividing then multiplying by ρ
+  linearized_air_pressure(ρ, state.ρe * invρ, gravitational_potential(orientation, aux), PhasePartition(TS))
+end
```
Comment (jkozdon) [inline src/Atmos/Model/linear.jl]: Is this actually linear?

Comment (simonbyrne) [reply src/Atmos/Model/linear.jl]: I'm not sure? Maybe we should just remove it for now, since it isn't being used.

File: test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+end
+
+# -------------- Problem constants ------------------- # 
+const (xmin,xmax)      = (0,1000)
+const (ymin,ymax)      = (0,400)
+const (zmin,zmax)      = (0,1000)
+const Ne        = (10,2,10)
+const polynomialorder = 4
+const dim       = 3
+const dt        = 0.01
```
Comment (jkozdon) [inline test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl]: This seems to be the same `dt` as the current explicit version of the [rising thermal bubble](https://github.com/climate-machine/CLIMA/blob/9594889aa4b0b461122bc08cc1ee73ccd0eba27c/test/DGmethods/compressible_Navier_Stokes/rising_bubble-model.jl#L47). Can this be made any bigger? (Also don't know if the explicit `dt` is as large as possible.)

File: src/Atmos/Model/linear.jl [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+vars_diffusive(lm::AtmosLinearModel, T) = @vars()
+vars_aux(lm::AtmosLinearModel, T) = vars_aux(lm.atmos,T)
+vars_integrals(lm::AtmosLinearModel,T) = @vars()
+
+
+update_aux!(dg::DGModel, lm::AtmosLinearModel, Q::MPIStateArray, auxstate::MPIStateArray, t::Real) = nothing
+integrate_aux!(lm::AtmosLinearModel, integ::Vars, state::Vars, aux::Vars) = nothing
+flux_diffusive!(lm::AtmosLinearModel, flux::Grad, state::Vars, diffusive::Vars, aux::Vars, t::Real) = nothing
+function wavespeed(lm::AtmosLinearModel, nM, state::Vars, aux::Vars, t::Real)
+  return soundspeed_air(Float64(T_0))
```
Comment (jkozdon) [inline src/Atmos/Model/linear.jl]: I would expect this to be the reference profile temperature here? @fxgiraldo or @mwarusz

File: src/Atmos/Model/linear.jl [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+end
+
+boundary_state!(nf, lm::AtmosLinearModel, x...) = nothing
```
Comment (jkozdon) [inline src/Atmos/Model/linear.jl]: Shouldn't we need boundary conditions for the linear problem? @fxgiraldo or @mwarusz

File: src/Atmos/Model/linear.jl (hunk 4 above)
Comment (simonbyrne) [reply src/Atmos/Model/linear.jl]: @charleskawczynski what function should I use here to get temperature of reference state?

File: src/Atmos/Model/linear.jl (hunk 2 above)
Comment (jkozdon) [reply src/Atmos/Model/linear.jl]: Also this is never actually used. Is that deliberate?

**Comment by jkozdon** [comment]
Also need to add this to the slurmci tests

File: src/Atmos/Model/linear.jl (hunk 4 above)
Comment (jkozdon) [reply src/Atmos/Model/linear.jl]: Don't know that it should be, just wondering.

Comment (simonbyrne) [reply src/Atmos/Model/linear.jl]: 🤦‍♂  that's how we specify the reference state.

File: test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl (hunk 3 above)
Comment (simonbyrne) [reply test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl]: We need a proper integration test for this, see #462.

File: src/Atmos/Model/linear.jl (hunk 4 above)
Comment (tapios) [reply src/Atmos/Model/linear.jl]: Let‘s avoid `T_0` for reference state temperature (for dynamical calculations) because we use `T_0` as the thermodynamic reference temperature (a PlanetParameter) that specifies at which temperature reference specific heats are evaluated.

Comment (tapios) [reply src/Atmos/Model/linear.jl]: The reference temperature is a function of height generally, and so would be the speed of sound. Do you want to take the `max` of this here?

Comment (charleskawczynski) [reply src/Atmos/Model/linear.jl]: Yeah, so we'll want something like `return soundspeed_air(aux.ref_state.T)`

Comment (simonbyrne) [reply src/Atmos/Model/linear.jl]: We store the reference temperature, so can just use that.

File: src/Atmos/Model/linear.jl [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  DF = eltype(state)
+  ref = aux.ref_state
+  e_pot = gravitational_potential(lm.atmos.orientation, aux)
+
+  flux.ρ = state.ρu
+  # TODO: use MoistThermodynamics.linearized_air_pressure 
+  # need to avoid dividing then multiplying by ρ
+  pL = state.ρ * DF(R_d) * DF(T_0) + DF(R_d) / DF(cv_d) * (state.ρe - state.ρ * e_pot)
+  flux.ρu += pL*I
+  flux.ρe = ((ref.ρe + ref.p)/ref.ρ - e_pot)*state.ρu
```
[the remark this answers is not in the document]
Comment (simonbyrne) [reply src/Atmos/Model/linear.jl]: this is my interpretation of eqn 2.35 in the overleaf docs: is this correct @tapios?

File: src/Atmos/Model/linear.jl (hunk 5 above)
Comment (tapios) [reply src/Atmos/Model/linear.jl]: Yes, you need linear boundary conditions. But keep in mind that the wave modes resolved here transport nothing. Hence, the boundary conditions are not the same as for the transport term. Just use linear reflecting boundary conditions.

File: src/Atmos/Model/linear.jl [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+end
+function flux_nondiffusive!(lm::AtmosAcousticLinearModel, flux::Grad, state::Vars, aux::Vars, t::Real)
+  DF = eltype(state)
+  ref = aux.ref_state
+  e_pot = gravitational_potential(lm.atmos.orientation, aux)
+
+  flux.ρ = state.ρu
+  # TODO: use MoistThermodynamics.linearized_air_pressure 
+  # need to avoid dividing then multiplying by ρ
+  pL = state.ρ * DF(R_d) * DF(T_0) + DF(R_d) / DF(cv_d) * (state.ρe - state.ρ * e_pot)
```
Comment (tapios) [inline src/Atmos/Model/linear.jl]: @charleskawczynski provided a function `linearized_air_pressure` in `MoistThermodynamics` for this. It seems preferable to me to use that.

File: src/Atmos/Model/linear.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (tapios) [reply src/Atmos/Model/linear.jl]: Correct, you want to subtract the potential energy if gravity waves are excluded. To include gravity waves, you leave the potential energy term here and in the momentum equations.

File: src/Atmos/Model/linear.jl [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+struct AtmosAcousticGravityLinearModel{M} <: AtmosLinearModel
+  atmos::M
+end
+function flux_nondiffusive!(lm::AtmosAcousticGravityLinearModel, flux::Grad, state::Vars, aux::Vars, t::Real)
+  DF = eltype(state)
+  ref = aux.ref_state
+  e_pot = gravitational_potential(lm.atmos.orientation, aux)
+
+  flux.ρ = state.ρu
+  pL = state.ρ * DF(R_d) * DF(T_0) + DF(R_d) / DF(cv_d) * (state.ρe - state.ρ * e_pot)
```
Comment (tapios) [inline src/Atmos/Model/linear.jl]: As above, use `linearized_air_pressure`?

File: src/Atmos/Model/linear.jl [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+vars_aux(lm::AtmosLinearModel, T) = vars_aux(lm.atmos,T)
+vars_integrals(lm::AtmosLinearModel,T) = @vars()
+
+
+update_aux!(dg::DGModel, lm::AtmosLinearModel, Q::MPIStateArray, auxstate::MPIStateArray, t::Real) = nothing
+integrate_aux!(lm::AtmosLinearModel, integ::Vars, state::Vars, aux::Vars) = nothing
+flux_diffusive!(lm::AtmosLinearModel, flux::Grad, state::Vars, diffusive::Vars, aux::Vars, t::Real) = nothing
+function wavespeed(lm::AtmosLinearModel, nM, state::Vars, aux::Vars, t::Real)
+  ref = aux.ref_state
+  return soundspeed_air(ref_state.T)
```
Comment (mwarusz) [inline src/Atmos/Model/linear.jl]:
```suggestion
  return soundspeed_air(ref.T)
```

File: test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+  linmodel = LinearType(model)
+  lindg = DGModel(linmodel,
+               grid,
+               Rusanov(),
+               CentralNumericalFluxDiffusive(),
+               CentralGradPenalty(); auxstate=dg.auxstate)
+
+  Q = init_ode_state(dg, DF(0))
+
+  linearsolver = GeneralizedMinimalResidual(10, Q, 1e-10)
```
Comment (mwarusz) [inline test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl]: `1e-10` is too high tolerance for `Float32` . Something like `1e-5` may work in that case.

File: test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl (hunk 3 above)
Comment (mwarusz) [reply test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl]: After adding boundary conditions for the linear model and fixing small problems mentioned in my recent comments I was able to run this test with much bigger time step (e.g. `dt = 0.4` worked).

Comment (jkozdon) [reply test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl]: Do you know the largest time step running in fully explicit mode?

Comment (fxgiraldo) [reply test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl]: It would be helpful to be able to be able to extract Courant number values (e.g., both with respect to acoustic waves and velocity). It would also be beneficial to subdivide this information into vertically and horizontally propagating waves.

File: src/Atmos/Model/linear.jl (hunk 7 above)
Comment (simonbyrne) [reply src/Atmos/Model/linear.jl]: Yes, see the comment. There were some issues in that it divides then multiplying by ρ.

File: test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl (hunk 10 above)
Comment (simonbyrne) [reply test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl]: How about?
```suggestion
  linearsolver = GeneralizedMinimalResidual(10, Q, sqrt(eps(DF)))
```

Comment (mwarusz) [reply test/DGmethods/compressible_Navier_Stokes/rising_bubble-model-imex.jl]: Seems fine. The tolerance in the old IMEX examples was set very high to get reproducible test results. In normal runs it will be much smaller. Whatever works to have a test that will not break on different machines is fine here.

File: src/Atmos/Model/linear.jl [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  ref = aux.ref_state
+  return soundspeed_air(ref.T)
+end
+
+function boundary_state!(nf::Rusanov, lm::AtmosLinearModel, x...)
+  atmos_boundary_state!(nf, NoFluxBC(), lm.atmos, x...)
+end
+function boundary_state!(nf::CentralNumericalFluxDiffusive, lm::AtmosLinearModel, x...)
+  nothing
+end
```
Comment (simonbyrne) [inline src/Atmos/Model/linear.jl]: @jkozdon @mwarusz does this seem a reasonable default?

**Comment by simonbyrne** [comment]
bors try

File: src/Atmos/Model/linear.jl (hunk 11 above)
Comment (jkozdon) [reply src/Atmos/Model/linear.jl]: Seems reasonable to me.

Comment (mwarusz) [reply src/Atmos/Model/linear.jl]: I agree, seems fine.

File: src/Atmos/Model/linear.jl (hunk 7 above)
Comment (charleskawczynski) [reply src/Atmos/Model/linear.jl]: I'll add a function in MoistThermo that doesn't require the multiply-divide
